// gist-box/lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var request_1 = __importDefault(require("@octokit/request"));
/**
 * The maximum number of characters in a pinned Gist line
 */
exports.MAX_LENGTH = 63;
/**
 * The maximum number of lines in a pinned Gist
 */
exports.MAX_LINES = 5;
/**
 * The maximum width of a pinned Gist image
 */
exports.MAX_WIDTH = 325;
/**
 * The maximum height of a pinned Gist image
 */
exports.MAX_HEIGHT = 100;
var GistBox = /** @class */ (function () {
    function GistBox(_a) {
        var id = _a.id, token = _a.token;
        this.id = id;
        this.token = token;
    }
    /**
     * Get the Gist
     */
    GistBox.prototype.get = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, request_1.default('GET /gists/:gist_id', {
                        gist_id: this.id,
                        headers: {
                            authorization: "token " + this.token
                        }
                    })];
            });
        });
    };
    /**
     * Update the Gist
     */
    GistBox.prototype.update = function (updates) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, gist, filename;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.get()];
                    case 1:
                        gist = _b.sent();
                        filename = Object.keys(gist.data.files)[0];
                        return [2 /*return*/, request_1.default('PATCH /gists/:gist_id', {
                                files: (_a = {}, _a[filename] = updates, _a),
                                gist_id: this.id,
                                headers: {
                                    authorization: "token " + this.token
                                }
                            })];
                }
            });
        });
    };
    return GistBox;
}());
exports.GistBox = GistBox;
